numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1120,6 +1120,30 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
         goto fail;
     }
 
+    if (!PyArray_CanCastArrayTo(indices, dtype, NPY_SAFE_CASTING)) {
+        if (PyArray_SIZE(indices) == 0) {
+            /*
+             * Unfortunately, if the input array is empty, then indices would
+             * likely have a default dtype of float64, which isn't compatibile.
+             * Force a recast of an empty array.
+             */
+            PyArrayObject *indices_tmp = NULL;
+            Py_INCREF(dtype);
+            indices_tmp = (PyArrayObject*)PyArray_FromArray(indices, dtype,
+                                                            NPY_ARRAY_FORCECAST);
```

Comment:
So you opted to cast empty arrays of incorrect type :). I will just mention again that I have a slight dislike, and thus be lazy and leave it someone else to merge ;).